domain/secretbackend/service/modelsecretbackendservice.go

```diff
@@ -58,27 +65,28 @@ func (s *ModelSecretBackendService) SetModelSecretBackend(ctx context.Context, b
 	if backendName == provider.Internal || backendName == kubernetes.BackendName {
 		return fmt.Errorf("secret backend name %q not valid%w", backendName, errors.Hide(secretbackenderrors.NotValid))
 	}
-
-	modelSecretBackend, err := s.st.GetModelSecretBackendDetails(ctx, s.modelID)
-	if err != nil {
-		return fmt.Errorf("cannot get model secret backend detail for %q: %w", s.modelID, err)
-	}
-	if backendName == provider.Auto {
-		switch modelSecretBackend.ModelType {
-		case coremodel.IAAS:
-			backendName = provider.Internal
-		case coremodel.CAAS:
-			backendName = kubernetes.BackendName
-		default:
-			// Should never happen.
-			return fmt.Errorf("cannot set model secret backend for unsupported model type %q for model %q",
-				modelSecretBackend.ModelType, s.modelID,
-			)
+	return s.st.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		modelSecretBackend, err := s.st.GetModelSecretBackendDetails(ctx, s.modelID)
+		if err != nil {
+			return fmt.Errorf("cannot get model secret backend detail for %q: %w", s.modelID, err)
 		}
-	}
-	err = s.st.SetModelSecretBackend(ctx, s.modelID, backendName)
-	if err != nil {
-		return fmt.Errorf("cannot set model secret backend for %q: %w", s.modelID, err)
-	}
-	return errors.Trace(err)
+		if backendName == provider.Auto {
+			switch modelSecretBackend.ModelType {
+			case coremodel.IAAS:
+				backendName = provider.Internal
+			case coremodel.CAAS:
+				backendName = kubernetes.BackendName
+			default:
+				// Should never happen.
+				return fmt.Errorf("cannot set model secret backend for unsupported model type %q for model %q",
+					modelSecretBackend.ModelType, s.modelID,
+				)
+			}
+		}
+		err = s.st.SetModelSecretBackend(ctx, s.modelID, backendName)
+		if err != nil {
+			return fmt.Errorf("cannot set model secret backend for %q: %w", s.modelID, err)
```

Comment:
As a drive by we should change these to avoid using "cannot".
"setting model secret backend for ..."